go/vt/vtgate/router.go

```diff
@@ -68,135 +68,50 @@ func (rtr *Router) Execute(ctx context.Context, sql string, bindVars map[string]
-	result := &sqltypes.Result{}
-	if len(lresult.Rows) == 0 && wantFields {
-		for k := range join.Vars {
-			vcursor.JoinVars[k] = nil
-		}
-		rresult, err := rtr.getFields(vcursor, join.Right)
-		if err != nil {
-			return nil, err
-		}
-		result.Fields = joinFields(lresult.Fields, rresult.Fields, join.Cols)
-		return result, nil
-	}
-	for _, lrow := range lresult.Rows {
-		for k, col := range join.Vars {
-			vcursor.JoinVars[k] = lrow[col]
-		}
-		rresult, err := rtr.execInstruction(vcursor, join.Right, wantFields)
-		if err != nil {
-			return nil, err
-		}
-		if wantFields {
-			wantFields = false
-			result.Fields = joinFields(lresult.Fields, rresult.Fields, join.Cols)
-		}
-		for _, rrow := range rresult.Rows {
-			result.Rows = append(result.Rows, joinRows(lrow, rrow, join.Cols))
-		}
-		if join.IsLeft && len(rresult.Rows) == 0 {
-			result.Rows = append(result.Rows, joinRows(lrow, nil, join.Cols))
-			result.RowsAffected++
-		} else {
-			result.RowsAffected += uint64(len(rresult.Rows))
-		}
-	}
-	return result, nil
-}
-
-func copyBindVars(bindVars map[string]interface{}) map[string]interface{} {
-	out := make(map[string]interface{})
-	for k, v := range bindVars {
-		out[k] = v
-	}
-	return out
-}
-
-func joinFields(lfields, rfields []*querypb.Field, cols []int) []*querypb.Field {
-	fields := make([]*querypb.Field, len(cols))
-	for i, index := range cols {
-		if index < 0 {
-			fields[i] = lfields[-index-1]
-			continue
-		}
-		fields[i] = rfields[index-1]
+// StreamExecute executes a streaming query.
+func (rtr *Router) StreamExecute(ctx context.Context, sql string, bindVars map[string]interface{}, tabletType topodatapb.TabletType, sendReply func(*sqltypes.Result) error) error {
+	if bindVars == nil {
+		bindVars = make(map[string]interface{})
 	}
-	return fields
-}
-
-func joinRows(lrow, rrow []sqltypes.Value, cols []int) []sqltypes.Value {
-	row := make([]sqltypes.Value, len(cols))
-	for i, index := range cols {
-		if index < 0 {
-			row[i] = lrow[-index-1]
-			continue
-		}
-		// rrow can be nil on left joins
-		if rrow != nil {
-			row[i] = rrow[index-1]
-		}
+	vcursor := newRequestContext(ctx, sql, bindVars, tabletType, nil, false, rtr)
+	plan, err := rtr.planner.GetPlan(sql)
+	if err != nil {
+		return err
 	}
-	return row
+	return plan.Instructions.StreamExec(vcursor, make(map[string]interface{}), true, sendReply)
 }
 
-// execRoute executes the route query for all route opcodes.
-func (rtr *Router) execRoute(vcursor *requestContext, route *planbuilder.Route) (*sqltypes.Result, error) {
+// ExecRoute executes the route query for all route opcodes.
+func (rtr *Router) ExecRoute(vcursor *requestContext, route *engine.Route, joinvars map[string]interface{}) (*sqltypes.Result, error) {
 	saved := copyBindVars(vcursor.bindVars)
 	defer func() { vcursor.bindVars = saved }()
-	for k := range route.JoinVars {
-		vcursor.bindVars[k] = vcursor.JoinVars[k]
+	for k := range joinvars {
+		vcursor.bindVars[k] = joinvars[k]
```

Comment:
Nit here:, would this be more efficient:

for k, v := range joinvars {
  vcursor.bindVars[k] = v
}

?